### PR TITLE
Reimplement DenseLayout as a Parallel Algorithm in Rust

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit re-implements the core of the DenseLayout transpiler pass
in Rust and to run in multiple threads. Previously this algorithm used
scipy sparse matrices to create a CSR sparse matrix representation of
the coupling graph and iterate over that to find a densely connected
subgraph in the coupling graph. This performs and scales well for modest
sized circuits and coupling graphs. But as the size of the coupling
graphs and circuits grows running the algorithm by iterating over the
sparse matrix in Python starts to hit a limit. The underlying traversal
can be efficiently executed in parallel using rust as the algorithm
iterates over the coupling graph in BFS order for each node to try and
find the best subgraph. We can do the BFS traversals in parallel and
then iteratively compare the results in parallel until the best is
found. This greatly speeds up the execution of the pass, for example
running on a 1081 qubit quantum volume circuit on 1081 qubit
heavy hexagon coupling graph takes ~135 seconds with the previous
version and ~0.14 seconds after this commit (on my local workstation
with 32 physical cores and 64 logical cores, scaling likely won't be as
good on smaller systems).

The tradeoff here comes in slightly increased memory consumption as
to have a shared representation of the adjacency matrix (and the error matrix)
between Python and Rust we use numpy arrays as they can be passed by
reference between the languages. In practice this will not matter much
until the graphs get truly large (e.g. to represent a 10000 qubit adjacency
matrix and error matrix would require ~1.6 GB of memory) and if it does
become an issue (either for memory or runtime performance) we can add a
shared compressed sparse matrix representation to Qiskit for use in both
Python in Rust.

### Details and comments

- [x] Update 4 dense layout tests for slightly different output
- [x] Add release notes
- [x] Add rust docstrings
- [x] Additional benchmarking and tuning

### CODE REVIEW

**Comment by ewinston** [comment]
Here is the scaling performance I saw for square random circuits run on 16 cores;

image

For the non-parallel version;

image

The parallel version is a couple of orders of magnitude faster, however the scaling is a bit worse which appears to agrees with with the down turn in the ratio of performance you saw.

File: src/dense_layout.rs
Comment (mtreinish) [inline src/dense_layout.rs:111]: I just remembered that we should probably switch here like we do in stochastic swap based on the env variables. I extract that in #7702 into a shared function. So we probably should wait for that PR to merge and I can use the `eval_parallel_env()` function added there in this PR to switch between parallel and single threaded versions here.

File: src/dense_layout.rs
Comment (kevinhartman) [inline src/dense_layout.rs:153]: Does this imply that the `cx_sum` will be double what it actually should be if the user passed a symmetric coupling map? Or maybe that the subgraph will end up with half as many edges?

File: test/python/transpiler/test_preset_passmanagers.py
Comment (kevinhartman) [inline test/python/transpiler/test_preset_passmanagers.py:632]: What makes this end up different? Presumably it's still deterministic?

File: releasenotes/notes/rust-denselayout-bc0f08874ad778d6.yaml
Comment (kevinhartman) [inline releasenotes/notes/rust-denselayout-bc0f08874ad778d6.yaml:5]:
```suggestion
    greatly improves the runtime performance of the pass. By default, it will
```

File: releasenotes/notes/rust-denselayout-bc0f08874ad778d6.yaml
Comment (kevinhartman) [inline releasenotes/notes/rust-denselayout-bc0f08874ad778d6.yaml:6]:
```suggestion
    use the number of logical CPUs on your local system, but you can control
```

File: releasenotes/notes/rust-denselayout-bc0f08874ad778d6.yaml
Comment (kevinhartman) [inline releasenotes/notes/rust-denselayout-bc0f08874ad778d6.yaml:7]:
```suggestion
    the number of threads used by the pass by setting the
```

File: releasenotes/notes/rust-denselayout-bc0f08874ad778d6.yaml
Comment (kevinhartman) [inline releasenotes/notes/rust-denselayout-bc0f08874ad778d6.yaml:9]:
```suggestion
    example, setting ``RAYON_NUM_THREADS=4`` will run the
```

File: releasenotes/notes/rust-denselayout-bc0f08874ad778d6.yaml
Comment (kevinhartman) [inline releasenotes/notes/rust-denselayout-bc0f08874ad778d6.yaml:10]:
```suggestion
    :class:`~.DenseLayout` pass with 4 threads.
```

File: qiskit/transpiler/passes/layout/dense_layout.py
Comment (kevinhartman) [inline qiskit/transpiler/passes/layout/dense_layout.py:65]: Can you clarify what this loop is doing?

Comment (mtreinish) [reply qiskit/transpiler/passes/layout/dense_layout.py:65]: Ah, the comment here is for the if statement below specifically (when faulty qubits are set, the `qubits in properties` != set of qubits` in coupling map). So it's just bailing if we have more qubits to read from the noise properties than are in the coupling map.

This loop is to get the measurement error for each qubit in the coupling map, in `BackendV1` (and `BaseBackend`) this is contained in the `qubits` field of `BackendProperties` object under the dict where `obj["name"] == "readout_error"`. It's a bit more intuitive in `BackendV2` with the `Target`, but I was saving that for a follow on PR.

File: test/python/transpiler/test_preset_passmanagers.py
Comment (mtreinish) [reply test/python/transpiler/test_preset_passmanagers.py:632]: It's actually caused by a bug in the python implementation before this. The noise properties were never actually populated in the error matrix for 2q errors as there was a type mismatch. It was trying to do `[0, 1] == (0, 1)` to see if an edge in the coupling map matched an edge in the noise properties, which was evaluating as False because of list vs tuple. I caught that and fixed it as part of this, so this is the change in result due to actually using the cx error rate for the device to pick the best dense layout. You can look at https://github.com/Qiskit/qiskit-terra/pull/7740/commits/7537709a9306aa5ffb17404a75dba7f4d47b3599 where I fixed this and I had to the update the test after making that change in: https://github.com/Qiskit/qiskit-terra/pull/7740/commits/34aeb1383d9488d942182c16d008d889dbb07a36 

This should always be deterministic for this test now (since the noise properties for the `FakeTokyo` backend are fixed).

File: src/dense_layout.rs
Comment (mtreinish) [reply src/dense_layout.rs:154]: Yeah, it implies it will be double, with a symmetric coupling map we'll have bidirectional cx gates so we'll have duplicate entries in the subgraph for each cx gate between any 2 qubits, this is just making it so we're only counting a single cx error rate between 2 qubits. That being said I did just copy this formula from the previous implementation. I think as part of #7113 I think we will probably have to refactor this part a bit and construct the error matrix differently.

File: src/dense_layout.rs
Comment (mtreinish) [reply src/dense_layout.rs:84]: Yeah, that's what it means here. In the physical sense it means all the cx gates in the coupling map are bidirectional. I made this a standalone parameter here because the coupling graph can cache this (and there is a retworkx native function to compute it quickly). This just influences how we define the error rate/score of a layout, if this is true we divide the cx error rate in half because there will be 2 for each pair of nodes in the graph.

File: src/dense_layout.rs
Comment (mtreinish) [reply src/dense_layout.rs:124]: Ah, yeah that's a good point I was a bit too concerned with the mechanics of porting from the sparse array access patterns and loops to a dense array in rust that I didn't really think this through. Making this just call `filter()` will be a lot simpler.

File: src/dense_layout.rs
Comment (mtreinish) [reply src/dense_layout.rs]: Done in: https://github.com/Qiskit/qiskit-terra/pull/7740/commits/a1d607deec2078d65ea14b1a14860a7626813ce3

File: src/dense_layout.rs
Comment (mtreinish) [reply src/dense_layout.rs]: Done in: https://github.com/Qiskit/qiskit-terra/pull/7740/commits/a1d607deec2078d65ea14b1a14860a7626813ce3

Comment (mtreinish) [reply src/dense_layout.rs]: Ah, actually I forgot to do the map, let me add that now

**Comment by mtreinish** [comment]
After the most recent changes based on the suggestions by @kevinhartman the overall performance improved quite substantially, although it looks like we still hit a scaling wall at roughly the same point:

32core_linear

**Comment by mtreinish** [comment]
I took a quick look at the profile of a 3000 qubit example and am not seeing an obvious hot spot. In parallel most of the work is in the parallel closure context and when I modified the iterator to be serially executed (changing `into_par_iter()` into `into_iter()` and adjusting things to compile) the perf profile just shows that ~95% of the time is in `qiskit_accelerate::dense_layout::best_subset{{closure}}` and everything below that is opaque. Which isn't super helpful

**Comment by mtreinish** [comment]
Thinking about this a bit more and talking to @georgios-ts I think the next step for this to address the scaling problems is to potentially add an implementation of this function to retworkx and leverage the better and more efficient graph representation retworkx uses (the scaling limits seem mostly around doing the bfs on the adjacency matrix). The coupling graph is already a retworkx graph internally so we could write this same basic algorithm directly on the graph data structure. Ideally we'd be able to do this in `qiskit._accelerate` directly and call retworkx's graph directly because both are rust. But unfortunately we're not able to share PyO3/Rust python objects between rust libraries (see: https://github.com/PyO3/pyo3/issues/1444 for more details) so working directly off of the retworkx graph isn't an option yet.

As an intermediate step we could rewrite this algorithm using petgraph (which is the upstream library that retworkx uses for the graph representation) in `qiskit._accelerate`. But for this PR I'm quite happy with this implementation as a first step, it provides a huge speed up with minimal changes to the code (at the cost of increased memory consumption) and a petgrah implementation would be a larger change. My thinking is we can leave either rewriting this in petgraph or adding a function to retworkx as a follow on step after this initial implementation merges.

File: src/dense_layout.rs
Comment (kevinhartman) [reply src/dense_layout.rs:155]: Sounds good. I'm wondering if `subgraph` itself would have twice as many items if this were a symmetric coupling map anyway, so I wonder if dividing by 2 is actually necessary (and if it's not, it'd make the result wrong).

File: src/dense_layout.rs
Comment (mtreinish) [reply src/dense_layout.rs]: Done in: https://github.com/Qiskit/qiskit-terra/pull/7740/commits/90860dc3a3bf3586a1851e409ded943a92d3409f

**Review by kevinhartman — APPROVED** [review]
LGTM.

File: src/dense_layout.rs
Comment (kevinhartman) [reply src/dense_layout.rs]: I think it'd be good to have a unit test to validate that `DenseLayout` works as expected wrt. error rates for symmetric coupling maps. Probably beyond the scope of the PR, since you haven't changed the logic.

Comment (mtreinish) [reply src/dense_layout.rs]: Yeah, I'll open an issue for it after this merges. The testing for this pass was a bit lacking so having more thorough testing would be a good thing in general.

**Comment by jakelishman** [comment]
I feel like we should at least acknowledge that Matthew called this branch `cthulhu_mckee` before it merges 🦑🐉

**Comment by mtreinish** [comment]
Hah, I'm glad someone noticed. I just kept misreading `reverse_cuthill_mckee()` as `reverse_cthulhu_mckee()` when I was developing this and named the branch as an inside joke to myself :)

**Pull request merged**
